Bluetooth: hci: spi: Use edge sensitive interrupt and mutex fixes.

HoZHel opened the pull request:
Revert to edge sensitive interrupt. (Fixes #59459 )
Optimized BlueNRG-MS access.
Addressed issue for mutex access of SPI.
Fixed chip select active state for x_nucleo_idb05a1.
I also tested samples/bluetooth/beacon on 96b_carbon board and it works fine; however, without my modifications in spi.c file the application crashes since ST boards do not support level trigger interrupt.

HoZHel commented:
@JordanYates I tried to fix the issue #59459 and tested on nucleo_l476rg with x_nucleo_idb05a1.
Everything looks good to me, but I did not test in any other platform. Could you please take a look and let me know about your opinion.
Moreover, would you please inform me regarding the platform you are using for testing changes.
This is my first pull request, and I set in draft state so that I can apply more changes.

JordanYates commented on drivers/bluetooth/hci/spi.c:
```diff
@@ -36,6 +36,7 @@ LOG_MODULE_REGISTER(bt_driver);
 /* Offsets */
 #define STATUS_HEADER_READY	0
 #define STATUS_HEADER_TOREAD	3
+#define STATUS_HEADER_TOWRITE	1
```
Where did this come from? Offset 1 in the HCI controller sample is `SANITY_CHECK`, nothing to do with writing.

JordanYates commented on boards/shields/x_nucleo_idb05a1/x_nucleo_idb05a1.overlay:
```diff
@@ -5,13 +5,14 @@
 
 &arduino_spi {
-	cs-gpios = <&arduino_header 1 (GPIO_ACTIVE_HIGH | GPIO_PULL_UP)>;	      /* A1 */
+	cs-gpios = <&arduino_header 1 (GPIO_ACTIVE_LOW | GPIO_PULL_UP)>;	      /* A1 */
```
Looks fine, but should be a separate commit in the pull request.

HoZHel replied:
I agree.

JordanYates commented on drivers/bluetooth/hci/spi.c:
```diff
@@ -302,36 +287,44 @@ static void bt_spi_rx_thread(void)
-		do {
+		/* On BlueNRG-MS, host is expected to read */
+		/* as long as IRQ pin is high */
+		while (irq_pin_high()) {
```
`irq_pin_high` is defined as `#define irq_pin_high(...) 0` when `!CONFIG_BT_SPI_BLUENRG`, so this code will now never run with other controllers.

HoZHel replied:
I think you are correct, and I will try to find another solution.

carlescufi commented on drivers/bluetooth/hci/spi.c:
```diff
@@ -302,36 +287,44 @@ static void bt_spi_rx_thread(void)
 		LOG_DBG("");
 
-		do {
+		/* On BlueNRG-MS, host is expected to read */
```
This is a common driver. Please do not refer to specific implementations here, or then encapsulate them inside a Kconfig option that does not refer to the particular SoC

JordanYates replied:
Can't fault @HoZHel for this, they've just moved this comment from elsewhere in the file :)
Personally I almost feel like the BlueNRG driver and the Zephyr driver should be separate implementations, its annoying to make the driver work for both.

JordanYates reviewed (changes requested):
Thanks for opening this PR and taking the time to try and fix this issue.
Can I please ask that you structure this PR as 3 commits:
1. revert my original commit that switched to level triggered interrupts
2. apply your proposed changes to the original state of the driver
3. apply your fix to the nucleo shield

This will make it easier to check the changes you are proposing, instead of having to mentally undo my original commit.

Personally I am using an in-house board that uses the in-tree [hci_spi](https://github.com/zephyrproject-rtos/zephyr/tree/main/samples/bluetooth/hci_spi) application on a nRF52832/nRF52840 as a Bluetooth controller to a nRF9160 main MCU.

I can't speak for the validity of the fix when using `CONFIG_BT_SPI_BLUENRG`, as that is not a controller that I have ever used or had access to.

HoZHel committed:
drivers: bluetooth: hci: spi: Revert to edge sensitive interrupt. (#59459)

Solved also a mutex issue with SPI access.

Signed-off-by: Ali Hozhabri <ali.hozhabri@st.com>

HoZHel commented on drivers/bluetooth/hci/spi.c:
```diff
@@ -302,36 +287,44 @@ static void bt_spi_rx_thread(void)
 	(void)memset(&txmsg, 0xFF, SPI_MAX_MSG_LEN);
-
+	/* Enable the interrupt line */
+	gpio_pin_interrupt_configure_dt(&irq_gpio, GPIO_INT_EDGE_TO_ACTIVE);
```
If the interrupt line is low, it will not be a problem since the interrupt is configured for rising edge.

JordanYates replied:
I should have said 'active' instead of low. If the line is already active at this point, the thread will block forever

HoZHel committed:
boards: shields: x_nucleo_idb05A1: Fix chip select active state

Set also controller data delay to zero, as it is not needed in BlueNRG-MS.

Signed-off-by: Ali Hozhabri <ali.hozhabri@st.com>

JordanYates commented:
How did you revert my original commit? The recommended way is `git revert <COMMIT>`. It looks like you've missed some changes I had there.

HoZHel commented:
> How did you revert my original commit? The recommended way is `git revert <COMMIT>`. It looks like you've missed some changes I had there.

I did a "soft" revert since I needed to add some other changes to keep the code functional, in particular SPI mutex fix. I also attached the diff file between 6b36e44887a695c0546dad2577cf383542a36123 (mine) and ca6c08f960d83265497af261c5eda2e782c47b26 (yours). Could you please check if you find something weird?
[diff.txt](https://github.com/zephyrproject-rtos/zephyr/files/12080315/diff.txt)

msmttchr commented:
> > How did you revert my original commit? The recommended way is `git revert <COMMIT>`. It looks like you've missed some changes I had there.
> 
> I did a "soft" revert since I needed to add some other changes to keep the code functional, in particular SPI mutex fix. I also attached the diff file between [6b36e44](https://github.com/zephyrproject-rtos/zephyr/commit/6b36e44887a695c0546dad2577cf383542a36123) (mine) and [ca6c08f](https://github.com/zephyrproject-rtos/zephyr/commit/ca6c08f960d83265497af261c5eda2e782c47b26) (yours). Could you please check if you find something weird? [diff.txt](https://github.com/zephyrproject-rtos/zephyr/files/12080315/diff.txt)

@JordanYates, for your convenience I have uploaded also the file result of `git revert -n ca6c08f960d83265497af261c5eda2e782c47b26` after fixing a conflict manually.
[spi.c.txt](https://github.com/zephyrproject-rtos/zephyr/files/12103515/spi.c.txt)

I think the changes of @HoZHel are correct. Also, when moving to edge sensitive, interrupts should be permanently enabled, so code to enable interrupt has been moved outside the main `while (true)` loop.
Possibly, interrupt enabling should be moved in `bt_spi_open` to make sure no edges are lost after device reset.

carlescufi commented:
@JordanYates can you please take another look?

JordanYates reviewed (changes requested):
The conflict also needs to be fixed

erwango reviewed (commented):
BlueNRG-MS is also enabled on following boards:
-disco_l475_iot1
-b_l4s5i_iot01a
-stm32l562e_dk
Can you check if they should also be updated ?

HoZHel commented:
> BlueNRG-MS is also enabled on following boards: -disco_l475_iot1 -b_l4s5i_iot01a -stm32l562e_dk Can you check if they should also be updated ?

@erwango, I do not have these boards available for testing, but from code inspection, I have noticed the following:

- The dts does not include the recently added parameter `controller-data-delay-us`, this should be set to zero to optimize performance, but it is not strictly needed.
- The chip select should be active low for stm32l562e_dk [(code fragment)](https://github.com/zephyrproject-rtos/zephyr/blob/d22e08c067a5df1f5dd215e3c4e2e8f41c4484d1/boards/arm/stm32l562e_dk/stm32l562e_dk_common.dtsi#L114C1-L115C1).

If you would like to have these fixed, I would think to do it in a separate PR. Do you agree ?
Do you have any chance to test the code on these platforms on your side ?

erwango commented:
> If you would like to have these fixed, I would think to do it in a separate PR. Do you agree ?

If not directly related to https://github.com/zephyrproject-rtos/zephyr/issues/59459, then it is fine to me.

> Do you have any chance to test the code on these platforms on your side ?

Yes, I'll be able to test.

HoZHel commented:
> > If you would like to have these fixed, I would think to do it in a separate PR. Do you agree ?
> 
> If not directly related to #59459, then it is fine to me.
> 
Indeed the proposed changes are not related to #59459. I will try to open another PR for them.
> > Do you have any chance to test the code on these platforms on your side ?
> 
> Yes, I'll be able to test.

Thanks for your availability for testing. I will also try to get these boards in my office.

HoZHel committed:
driver: bluetooth: hci: fix potential "block forever" issue

Interrupt is enabled before reset is released to make sure that the
first IRQ edge is captured, and rx thread can process it.

Remove delay in spi_open as it was redundant due to the "sem_initialised"
semaphore.

Signed-off-by: Ali Hozhabri <ali.hozhabri@st.com>

erwango commented:
Reassigning to @jhedberg as most impact is in bluetooth subsystem

erwango reviewed (approved):
Approved for shield part.

HoZHel commented:
> @HoZHel Can you fix issues reported in CI:

Done.
